Lib/pty.py

```diff
@@ -170,3 +175,128 @@ def spawn(argv, master_read=_read, stdin_read=_read):
 
     os.close(master_fd)
     return os.waitpid(pid, 0)[1]
+
+# Author: Soumendra Ganguly.
+SIGWINCH = signal.SIGWINCH
+
+def _login_pty(master_fd, slave_fd):
+    """Given a pty, makes the calling process a session leader, makes the pty slave
+    its controlling terminal, stdin, stdout, and stderr. Closes both pty ends."""
+    # Establish a new session.
+    os.setsid()
+    os.close(master_fd)
+
+    try:
+        fcntl.ioctl(slave_fd, termios.TIOCSCTTY) # Make the pty slave the controlling terminal.
+    except:
+        os.close(slave_fd)
+        raise
+
+    # Slave becomes stdin/stdout/stderr.
+    os.dup2(slave_fd, STDIN_FILENO)
+    os.dup2(slave_fd, STDOUT_FILENO)
+    os.dup2(slave_fd, STDERR_FILENO)
+    if (slave_fd > STDERR_FILENO):
+        os.close(slave_fd)
+
+def _winresz(pty_slave):
+    """Resize window."""
+    w = struct.pack('HHHH', 0, 0, 0, 0)
+    s = fcntl.ioctl(STDIN_FILENO, termios.TIOCGWINSZ, w)
+    fcntl.ioctl(pty_slave, termios.TIOCSWINSZ, s)
+
+def _create_hwinch(pty_slave):
+    """Creates SIGWINCH handler."""
+    def _hwinch(signum, frame):
+        try:
+            _winresz(pty_slave)
+        except:
+            pass
+    return _hwinch
+
+def _cleanup(master_fd, slave_fd, old_hwinch, tty_mode):
+    """Performs cleanup in wspawn."""
+
+    # Close both pty ends.
+    os.close(master_fd)
+    os.close(slave_fd)
+
+    # Restore original tty attributes.
+    if tty_mode != None:
+        tty.tcsetattr(STDIN_FILENO, tty.TCSAFLUSH, tty_mode)
+
+    # Restore original SIGWINCH signal handler.
+    try:
+        signal.signal(SIGWINCH, old_hwinch)
+    except:
+        pass
+
+def _ekill(child_pid):
+    """Kill spawned process due to exception."""
+    try:
+        os.kill(child_pid, signal.SIGTERM)
+        time.sleep(1)
+        os.kill(child_pid, signal.SIGKILL)
+    except:
+        pass
+    try:
+        os.waitpid(child_pid, 0)
+    except:
+        pass
+
+def _wcopy(master_fd, child_pid, master_read=_read, stdin_read=_read, timeout=0.01):
+    """Parent copy loop for wspawn."""
+    fds = [master_fd, STDIN_FILENO]
+    ret = (0,0)
+    while True:
+        rfds, wfds, xfds = select(fds, [], [], timeout)
+        if ret == (0,0):
+            ret = os.waitpid(child_pid, os.WNOHANG)
+        elif len(rfds) == 0:
+            break;
+        if master_fd in rfds:
+            data = master_read(master_fd)
+            if not data:  # Reached EOF.
+                fds.remove(master_fd)
+            else:
+                os.write(STDOUT_FILENO, data)
+        if STDIN_FILENO in rfds:
+            data = stdin_read(STDIN_FILENO)
+            if not data:
+                fds.remove(STDIN_FILENO)
+            else:
+                _writen(master_fd, data)
+    return ret
+
```

Comment:
"wspawn" is spawn+the following differences.

1. It sets window size at the beginning.
2. It registers a SIGWINCH handler.
3. It does not depend on OSError to return. It does a clean return instead [ see "_wcopy" above ].